src/request.rs

```diff
@@ -113,42 +119,76 @@ impl<'a> Request<'a> {
                     "INIT response: ABI {}.{}, flags {:#x}, max readahead {}, max write {}",
                     init.major, init.minor, init.flags, init.max_readahead, init.max_write
                 );
-                se.initialized = true;
-                reply.ok(&init);
+                se.initialized
+                    .store(true, std::sync::atomic::Ordering::Relaxed);
+                reply.ok(&init).await;
             }
             // Any operation is invalid before initialization
-            _ if !se.initialized => {
+            _ => {
                 warn!("Ignoring FUSE operation before init: {}", self.request);
-                self.reply::<ReplyEmpty>().error(EIO);
+                self.reply::<ReplyEmpty>(&ch).error(EIO).await;
             }
+        }
+    }
+
+    pub(crate) async fn maybe_destroy_dispatch(
```

Comment:
Does this need to be a separate method from `dispatch()`? I think you could change the return type of `dispatch()` to `Result<bool>` and merge them back into one method

Reply:
Your right -- i'll clean this up, this ties into the other comment below about mutability why the destroy on the filesystem is being called where it is. if we forego the mutability/ownership of the destroy method on the filesystem we can call it anywhere though. (the main dispatch loop without this no longer requires a session, but i believe i had this for mutability reasons. let me try collapse them back and pass through the session).